In [1]:
import warnings

warnings.filterwarnings("ignore")
# 26 remaining

# 100 pandas puzzles

> Inspired by 100 Numpy exerises, here are 100\* short puzzles for testing your knowledge of pandas' power.<br><br>
> Since pandas is a large library with many different specialist features and functions, these excercises focus mainly on the fundamentals of manipulating data (indexing, grouping, aggregating, cleaning), making use of the core DataFrame and Series objects.<br><br>
> Many of the excerises here are stright-forward in that the solutions require no more than a few lines of code (in pandas or NumPy... don't go using pure Python or Cython!). Choosing the right methods and following best practices is the underlying goal.<br><br>
> The exercises are loosely divided in sections. Each section has a difficulty rating; these ratings are subjective, of course, but should be a seen as a rough guide as to how inventive the required solution is.<br><br>
> If you're just starting out with pandas and you are looking for some other resources, the official documentation is very extensive. In particular, some good places get a broader overview of pandas are...<br><br>


# Importing Pandas


## Getting started and checking you pandas setup

### easy


1. Import pandas under the alias pd.


In [2]:
import pandas as pd

2. Print the version of pandas that has been imported.


In [3]:
pd.__version__

'2.2.1'

3. Print out all the version information of the libraries that are required by the pandas library.


In [4]:
pd.show_versions()


INSTALLED VERSIONS
------------------
commit                : bdc79c146c2e32f2cab629be240f01658cfb6cc2
python                : 3.12.2.final.0
python-bits           : 64
OS                    : Windows
OS-release            : 11
Version               : 10.0.22621
machine               : AMD64
processor             : Intel64 Family 6 Model 140 Stepping 1, GenuineIntel
byteorder             : little
LC_ALL                : None
LANG                  : None
LOCALE                : English_United States.1252

pandas                : 2.2.1
numpy                 : 1.26.4
pytz                  : 2024.1
dateutil              : 2.8.2
setuptools            : None
pip                   : 24.0
Cython                : None
pytest                : None
hypothesis            : None
sphinx                : None
blosc                 : None
feather               : None
xlsxwriter            : None
lxml.etree            : None
html5lib              : None
pymysql               : None
psycopg2           

---


## Dataframes basic

### A few of the fundamental routines for selecting, sorting, adding and aggregating data in DataFrames

#### easy


4. Create a DataFrame df from this dictionary data which has the index labels.


In [5]:
import numpy as np

data = {
    "animal": [
        "cat",
        "cat",
        "snake",
        "dog",
        "dog",
        "cat",
        "snake",
        "cat",
        "dog",
        "dog",
    ],
    "age": [2.5, 3, 0.5, np.nan, 5, 2, 4.5, np.nan, 7, 3],
    "visits": [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
    "priority": ["yes", "yes", "no", "yes", "no", "no", "no", "yes", "no", "no"],
}

labels = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j"]

data = pd.DataFrame(data=data, index=labels)

In [6]:
data

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no
d    dog  NaN       3      yes
e    dog  5.0       2       no
f    cat  2.0       3       no
g  snake  4.5       1       no
h    cat  NaN       1      yes
i    dog  7.0       2       no
j    dog  3.0       1       no

5. Display a summary of the basic information about this DataFrame and its data (hint: there is a single method that can be called on the DataFrame).


In [7]:
# way 1
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10 entries, a to j
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   animal    10 non-null     object 
 1   age       8 non-null      float64
 2   visits    10 non-null     int64  
 3   priority  10 non-null     object 
dtypes: float64(1), int64(1), object(2)
memory usage: 400.0+ bytes


In [8]:
# way 2
data.describe()

age     visits
count  8.000000  10.000000
mean   3.437500   1.900000
std    2.007797   0.875595
min    0.500000   1.000000
25%    2.375000   1.000000
50%    3.000000   2.000000
75%    4.625000   2.750000
max    7.000000   3.000000

6. Return the first 3 rows of the DataFrame df.


In [9]:
# way 1
data.head(3)

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no

In [10]:
# way 2
data.loc[:"c", :]

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no

In [11]:
# way 3
data.iloc[:3, :]

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no

7. Select just the 'animal' and 'age' columns from the DataFrame df.


In [12]:
# way 1
data[["animal", "age"]]

animal  age
a    cat  2.5
b    cat  3.0
c  snake  0.5
d    dog  NaN
e    dog  5.0
f    cat  2.0
g  snake  4.5
h    cat  NaN
i    dog  7.0
j    dog  3.0

In [13]:
# way 2
data.loc[:, ["animal", "age"]]

animal  age
a    cat  2.5
b    cat  3.0
c  snake  0.5
d    dog  NaN
e    dog  5.0
f    cat  2.0
g  snake  4.5
h    cat  NaN
i    dog  7.0
j    dog  3.0

In [14]:
# way 3
data.iloc[:, :2]

animal  age
a    cat  2.5
b    cat  3.0
c  snake  0.5
d    dog  NaN
e    dog  5.0
f    cat  2.0
g  snake  4.5
h    cat  NaN
i    dog  7.0
j    dog  3.0

8. Select the data in rows [3, 4, 8] and in columns ['animal', 'age'].


In [15]:
# way 1
data.iloc[[3, 4, 8], :2]

animal  age
d    dog  NaN
e    dog  5.0
i    dog  7.0

In [16]:
# way 2
data.loc[data.index[[3, 4, 8]], ["animal", "age"]]

animal  age
d    dog  NaN
e    dog  5.0
i    dog  7.0

9. Select only the rows where the number of visits is greater than 3.


In [17]:
# way 1
data[data["visits"] > 3]

Empty DataFrame
Columns: [animal, age, visits, priority]
Index: []

In [18]:
# way 2
data.loc[data["visits"] > 3]

Empty DataFrame
Columns: [animal, age, visits, priority]
Index: []

10. Select the rows where the age is missing, i.e. it is NaN.


In [19]:
data[data["age"].isna()]

animal  age  visits priority
d    dog  NaN       3      yes
h    cat  NaN       1      yes

11. Select the rows where the animal is a cat and the age is less than 3.


In [20]:
data[(data["animal"] == "cat") & (data["age"] < 3)]

animal  age  visits priority
a    cat  2.5       1      yes
f    cat  2.0       3       no

12. Select the rows the age is between 2 and 4 (inclusive).


In [21]:
# way 1
data[(data["age"] >= 2) & (data["age"] <= 4)]

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
f    cat  2.0       3       no
j    dog  3.0       1       no

In [22]:
# way 2
data[data["age"].between(2, 4)]

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
f    cat  2.0       3       no
j    dog  3.0       1       no

13. Change the age in row 'f' to 1.5.


In [23]:
data.loc["f", "age"] = 1.5

In [24]:
data

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no
d    dog  NaN       3      yes
e    dog  5.0       2       no
f    cat  1.5       3       no
g  snake  4.5       1       no
h    cat  NaN       1      yes
i    dog  7.0       2       no
j    dog  3.0       1       no

14. Calculate the sum of all visits in df (i.e. find the total number of visits).


In [25]:
data["visits"].sum()

19

15. Calculate the mean age for each different animal in df.


In [26]:
# way 1
data.groupby(["animal"])["age"].mean()

animal
cat      2.333333
dog      5.000000
snake    2.500000
Name: age, dtype: float64

In [27]:
# way 2
data.groupby("animal").agg({"age": "mean"})

age
animal          
cat     2.333333
dog     5.000000
snake   2.500000

16. Append a new row 'k' to df with your choice of values for each column. Then delete that row to return the original DataFrame.


In [28]:
# way 1
data.loc["k"] = ["eagle", 3, 3, "yes"]
data
data = data.drop(index="k")
data

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no
d    dog  NaN       3      yes
e    dog  5.0       2       no
f    cat  1.5       3       no
g  snake  4.5       1       no
h    cat  NaN       1      yes
i    dog  7.0       2       no
j    dog  3.0       1       no

17. Count the number of each type of animal in df.


In [29]:
# way 1
data["animal"].value_counts()

animal
cat      4
dog      4
snake    2
Name: count, dtype: int64

In [30]:
# way 2
data.groupby("animal").size()

animal
cat      4
dog      4
snake    2
dtype: int64

18. Sort df first by the values in the 'age' in decending order, then by the value in the 'visits' column in ascending order (so row i should be first, and row d should be last).


In [31]:
data.sort_values(by=["age", "visits"], ascending=[False, True])

animal  age  visits priority
i    dog  7.0       2       no
e    dog  5.0       2       no
g  snake  4.5       1       no
j    dog  3.0       1       no
b    cat  3.0       3      yes
a    cat  2.5       1      yes
f    cat  1.5       3       no
c  snake  0.5       2       no
h    cat  NaN       1      yes
d    dog  NaN       3      yes

19. The 'priority' column contains the values 'yes' and 'no'. Replace this column with a column of boolean values: 'yes' should be True and 'no' should be False.


In [32]:
data["priority"] = data["priority"].map({"yes": True, "no": False})

In [33]:
data

animal  age  visits  priority
a    cat  2.5       1      True
b    cat  3.0       3      True
c  snake  0.5       2     False
d    dog  NaN       3      True
e    dog  5.0       2     False
f    cat  1.5       3     False
g  snake  4.5       1     False
h    cat  NaN       1      True
i    dog  7.0       2     False
j    dog  3.0       1     False

20. In the 'animal' column, change the 'snake' entries to 'python'.


In [34]:
data["animal"] = data["animal"].replace({"snake": "python"})

In [35]:
data

animal  age  visits  priority
a     cat  2.5       1      True
b     cat  3.0       3      True
c  python  0.5       2     False
d     dog  NaN       3      True
e     dog  5.0       2     False
f     cat  1.5       3     False
g  python  4.5       1     False
h     cat  NaN       1      True
i     dog  7.0       2     False
j     dog  3.0       1     False

21. For each animal type and each number of visits, find the mean age. In other words, each row is an animal, each column is a number of visits and the values are the mean ages (hint: use a pivot table).


In [36]:
data

animal  age  visits  priority
a     cat  2.5       1      True
b     cat  3.0       3      True
c  python  0.5       2     False
d     dog  NaN       3      True
e     dog  5.0       2     False
f     cat  1.5       3     False
g  python  4.5       1     False
h     cat  NaN       1      True
i     dog  7.0       2     False
j     dog  3.0       1     False

In [37]:
data.pivot_table(index="animal", columns="visits", values="age", aggfunc="mean")

visits    1    2     3
animal                
cat     2.5  NaN  2.25
dog     3.0  6.0   NaN
python  4.5  0.5   NaN

---

## DataFrames: beyond the basics

### Slightly trickier: you may need to combine two or more methods to get the right answer

#### Difficulty: medium

> The previous section was tour through some basic but essential DataFrame operations. Below are some ways that you might need to cut your data, but for which there is no single "out of the box" method.


22. You have a DataFrame df with a column 'A' of integers.
    > For example: df = pd.DataFrame({'A': [1, 2, 2, 3, 4, 5, 5, 5, 6, 7, 7]})<br>

- How do you filter out rows which contain the same integer as the row immediately above?<br>
- You should be left with a column containing the following values:<br>
  1, 2, 3, 4, 5, 6, 7


In [38]:
df = pd.DataFrame({"A": [1, 2, 2, 3, 4, 5, 5, 5, 6, 7, 7]})

In [39]:
df[df["A"] == df["A"].shift()]

A
2   2
6   5
7   5
10  7

23. Given a DataFrame of numeric values, say<br>

> df = pd.DataFrame(np.random.random(size=(5, 3))) # a 5x3 frame of float values<br>

- how do you subtract the row mean from each element in the row?


In [40]:
df = pd.DataFrame(np.random.random(size=(5, 3)))

In [41]:
df

0         1         2
0  0.344029  0.627706  0.352806
1  0.283927  0.936144  0.670396
2  0.316569  0.915806  0.173875
3  0.096401  0.906943  0.802970
4  0.180200  0.736838  0.109413

In [42]:
df.mean(axis=1)

0    0.441514
1    0.630156
2    0.468750
3    0.602104
4    0.342150
dtype: float64

In [43]:
df.sub(df.mean(axis=1), axis=0)

0         1         2
0 -0.097485  0.186192 -0.088707
1 -0.346228  0.305988  0.040240
2 -0.152181  0.447056 -0.294875
3 -0.505704  0.304839  0.200865
4 -0.161950  0.394687 -0.232737

24. Suppose you have DataFrame with 10 columns of real numbers, for example:<br>
    > df = pd.DataFrame(np.random.random(size=(5, 10)), columns=list('abcdefghij'))<br>

- Which column of numbers has the smallest sum? Return that column's label.


In [44]:
df = pd.DataFrame(np.random.random(size=(5, 10)), columns=list("abcdefghij"))

In [45]:
df

a         b         c         d         e         f         g  \
0  0.114577  0.606608  0.566058  0.311412  0.632609  0.656823  0.103951   
1  0.804189  0.288711  0.585086  0.719573  0.310537  0.902315  0.046791   
2  0.155235  0.102208  0.786258  0.823379  0.900695  0.213411  0.700715   
3  0.143618  0.822537  0.407167  0.107066  0.453818  0.992619  0.315490   
4  0.793527  0.162252  0.586798  0.342731  0.882408  0.173780  0.167543   

          h         i         j  
0  0.356484  0.434874  0.888114  
1  0.864092  0.545854  0.185422  
2  0.021805  0.511495  0.579789  
3  0.774290  0.430936  0.344803  
4  0.323732  0.609898  0.789521

In [46]:
df.sum().idxmin()

'g'

25. How do you count how many unique rows a DataFrame has (i.e. ignore all rows that are duplicates)? As input, use a DataFrame of zeros and ones with 10 rows and 3 columns.

> df = pd.DataFrame(np.random.randint(0, 2, size=(10, 3)))


In [47]:
df = pd.DataFrame(np.random.randint(0, 2, size=(10, 3)))

In [48]:
df

0  1  2
0  1  1  1
1  0  0  1
2  0  0  1
3  1  1  0
4  1  0  1
5  1  1  0
6  1  1  0
7  1  1  1
8  1  1  1
9  1  1  1

In [49]:
df.drop_duplicates().shape[0]

4

26. In the cell below, you have a DataFrame df that consists of 10 columns of floating-point numbers. Exactly 5 entries in each row are NaN values.

- For each row of the DataFrame, find the column which contains the third NaN value.<br>
  > You should return a Series of column labels: e, c, d, h, d


In [50]:
nan = np.nan

data = [
    [0.04, nan, nan, 0.25, nan, 0.43, 0.71, 0.51, nan, nan],
    [nan, nan, nan, 0.04, 0.76, nan, nan, 0.67, 0.76, 0.16],
    [nan, nan, 0.5, nan, 0.31, 0.4, nan, nan, 0.24, 0.01],
    [0.49, nan, nan, 0.62, 0.73, 0.26, 0.85, nan, nan, nan],
    [nan, nan, 0.41, nan, 0.05, nan, 0.61, nan, 0.48, 0.68],
]

columns = list("abcdefghij")

df = pd.DataFrame(data, columns=columns)

In [51]:
df

a   b     c     d     e     f     g     h     i     j
0  0.04 NaN   NaN  0.25   NaN  0.43  0.71  0.51   NaN   NaN
1   NaN NaN   NaN  0.04  0.76   NaN   NaN  0.67  0.76  0.16
2   NaN NaN  0.50   NaN  0.31  0.40   NaN   NaN  0.24  0.01
3  0.49 NaN   NaN  0.62  0.73  0.26  0.85   NaN   NaN   NaN
4   NaN NaN  0.41   NaN  0.05   NaN  0.61   NaN  0.48  0.68

27. A DataFrame has a column of groups 'grps' and and column of integer values 'vals'

- df = pd.DataFrame({'grps': list('aaabbcaabcccbbc'),
  'vals': [12,345,3,1,45,14,4,52,54,23,235,21,57,3,87]}) <br>
- For each group, find the sum of the three greatest values. You should end up with the answer as follows:<br>
  > grps<br>
  > a 409<br>
  > b 156<br>
  > c 345<br>


In [52]:
df = pd.DataFrame(
    {
        "grps": list("aaabbcaabcccbbc"),
        "vals": [12, 345, 3, 1, 45, 14, 4, 52, 54, 23, 235, 21, 57, 3, 87],
    }
)

In [53]:
def calculate_sum_greatest_three(row):
    row = row.sort_values(ascending=False)
    return row.iloc[:3].sum()


df.groupby("grps").agg({"vals": calculate_sum_greatest_three})

vals
grps      
a      409
b      156
c      345

28. The DataFrame df constructed below has two integer columns 'A' and 'B'. The values in 'A' are between 1 and 100 (inclusive).<br>

- For each group of 10 consecutive integers in 'A' (i.e. (0, 10], (10, 20], ...), calculate the sum of the corresponding values in column 'B'.
- The answer should be a Series as follows:<br>
  > A<br>
  > (0, 10] 635<br>
  > (10, 20] 360<br>
  > (20, 30] 315<br>
  > (30, 40] 306<br>
  > (40, 50] 750<br>
  > (50, 60] 284<br>
  > (60, 70] 424<br>
  > (70, 80] 526<br>
  > (80, 90] 835<br>
  > (90, 100] 852


In [54]:
df = pd.DataFrame(
    np.random.RandomState(8765).randint(1, 101, size=(100, 2)), columns=["A", "B"]
)

In [55]:
df["cuts"] = pd.cut(x=df["A"], bins=[0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100])

In [56]:
df.groupby("cuts")["B"].sum()

cuts
(0, 10]      635
(10, 20]     360
(20, 30]     315
(30, 40]     306
(40, 50]     750
(50, 60]     284
(60, 70]     424
(70, 80]     526
(80, 90]     835
(90, 100]    852
Name: B, dtype: int32

---

## DataFrames: harder problems

### These might require a bit of thinking outside the box...

...but all are solvable using just the usual pandas/NumPy methods (and so avoid using explicit for loops)

#### Difficulty: hard


29. Consider a DataFrame df where there is an integer column 'X':
    > df = pd.DataFrame({'X': [7, 2, 0, 3, 4, 2, 5, 0, 3, 4]})

- For each value, count the difference back to the previous zero (or the start of the Series, whichever is closer). These values should therefore be<br>
  [1, 2, 0, 1, 2, 3, 4, 0, 1, 2]<br>
  Make this a new column 'Y'.


In [57]:
df = pd.DataFrame({"X": [7, 2, 0, 3, 4, 2, 5, 0, 3, 4]})
df

X
0  7
1  2
2  0
3  3
4  4
5  2
6  5
7  0
8  3
9  4

30. Consider the DataFrame constructed below which contains rows and columns of numerical data.

- Create a list of the column-row index locations of the 3 largest values in this DataFrame. In this case, the answer should be:<br>
  [(5, 7), (6, 4), (2, 5)]


In [58]:
df = pd.DataFrame(np.random.RandomState(30).randint(1, 101, size=(8, 8)))
df

0   1   2   3   4   5   6   7
0  38  38  46  46  13  24   3  54
1  18  47   4  42   8  66  50  46
2  62  36  19  19  77  17   7  63
3  28  47  46  65  63  12  16  24
4  14  51  34  56  29  59  92  79
5  58  76  96  45  38  76  58  40
6  10  34  48  40  37  23  41  26
7  55  70  91  27  79  92  20  31

31. You are given the DataFrame below with a column of group IDs, 'grps', and a column of corresponding integer values, 'vals'.

> df = pd.DataFrame({"vals": np.random.RandomState(31).randint(-30, 30, size=15),"grps": np.random.RandomState(31).choice(["A", "B"], 15)})

- Create a new column 'patched_values' which contains the same values as the 'vals' any negative values in 'vals' with the group mean:

|     | vals | grps | patched_vals |
| --: | ---: | :--- | :----------- |
|   0 |  -12 | A    | 13.6         |
|   1 |   -7 | B    | 28.0         |
|   2 |  -14 | A    | 13.6         |
|   3 |    4 | A    | 4.0          |
|   4 |   -7 | A    | 13.6         |
|   5 |   28 | B    | 28.0         |
|   6 |   -2 | A    | 13.6         |
|   7 |   -1 | A    | 13.6         |
|   8 |    8 | A    | 8.0          |
|   9 |   -2 | B    | 28.0         |
|  10 |   28 | A    | 28.0         |
|  11 |   12 | A    | 12.0         |
|  12 |   16 | A    | 16.0         |
|  13 |  -24 | A    | 13.6         |
|  14 |  -12 | A    | 13.6         |


In [59]:
df = pd.DataFrame({"vals": np.random.RandomState(31).randint(-30, 30, size=15),"grps": np.random.RandomState(31).choice(["A", "B"], 15)})
df

vals grps
0    -12    A
1     -7    B
2    -14    A
3      4    A
4     -7    A
5     28    B
6     -2    A
7     -1    A
8      8    A
9     -2    B
10    28    A
11    12    A
12    16    A
13   -24    A
14   -12    A

In [60]:
def create_patched_vals(row):
    if row['vals'] < 0:
        return df[(row['grps'] == df['grps']) & (df['vals'] > 0) ]['vals'].mean()
    return row['vals']
    
df['patched_vals'] = df.apply(create_patched_vals, axis=1)

In [61]:
df

vals grps  patched_vals
0    -12    A          13.6
1     -7    B          28.0
2    -14    A          13.6
3      4    A           4.0
4     -7    A          13.6
5     28    B          28.0
6     -2    A          13.6
7     -1    A          13.6
8      8    A           8.0
9     -2    B          28.0
10    28    A          28.0
11    12    A          12.0
12    16    A          16.0
13   -24    A          13.6
14   -12    A          13.6

32. Implement a rolling mean over groups with window size 3, which ignores NaN value. For example consider the following DataFrame:

> df = pd.DataFrame({'group': list('aabbabbbabab'),'value': [1, 2, 3, np.nan, 2, 3, np.nan, 1, 7, 3, np.nan, 8]})

|     | group | value |
| --: | :---- | ----: |
|   0 | a     |     1 |
|   1 | a     |     2 |
|   2 | b     |     3 |
|   3 | b     |   NaN |
|   4 | a     |     2 |
|   5 | b     |     3 |
|   6 | b     |   NaN |
|   7 | b     |     1 |
|   8 | a     |     7 |
|   9 | b     |     3 |
|  10 | a     |   NaN |
|  11 | b     |     8 |

- The goal is to compute the Series:

|     |        0 |
| --: | -------: |
|   0 | 1.000000 |
|   1 | 1.500000 |
|   2 | 3.000000 |
|   3 | 3.000000 |
|   4 | 1.666667 |
|   5 | 3.000000 |
|   6 | 3.000000 |
|   7 | 2.000000 |
|   8 | 3.666667 |
|   9 | 2.000000 |
|  10 | 4.500000 |
|  11 | 4.000000 |

- E.g. the first window of size three for group 'b' has values 3.0, NaN and 3.0 and occurs at row index 5. Instead of being NaN the value in the new column at this row index should be 3.0 (just the two non-NaN values are used to compute the mean (3+3)/2)


In [62]:
df = pd.DataFrame({'group': list('aabbabbbabab'),'value': [1, 2, 3, np.nan, 2, 3, np.nan, 1, 7, 3, np.nan, 8]})

---

## Series and DatetimeIndex

### Exercises for creating and manipulating Series with datetime data

#### Difficulty: easy/medium

- pandas is fantastic for working with dates and times. These puzzles explore some of this functionality.


33. Create a DatetimeIndex that contains each business day of 2015 and use it to index a Series of random numbers. Let's call this Series s.


34. Find the sum of the values in s for every Wednesday.


35. For each calendar month in s, find the mean of values.


36. For each group of four consecutive calendar months in s, find the date on which the highest value occurred.


37. Create a DateTimeIndex consisting of the third Thursday in each month for the years 2015 and 2016.


## Cleaning Data

### Making a DataFrame easier to work with

#### Difficulty: easy/medium

- It happens all the time: someone gives you data containing malformed strings, Python, lists and missing data. How do you tidy it up so you can get on with the analysis?

- Take this monstrosity as the DataFrame to use in the following puzzles:

> df = pd.DataFrame({'From_To': ['LoNDon_paris', 'MAdrid_miLAN', 'londON_StockhOlm','Budapest_PaRis', 'Brussels_londOn'],
'FlightNumber': [10045, np.nan, 10065, np.nan, 10085],'RecentDelays': [[23, 47], [], [24, 43, 87], [13], [67, 32]],'Airline': ['KLM(!)', '<Air France> (12)', '(British Airways. )','12. Air France', '"Swiss Air"']})

- Formatted, it looks like this:

|     | From_To          | FlightNumber | RecentDelays | Airline             |
| --: | :--------------- | -----------: | :----------- | :------------------ |
|   0 | LoNDon_paris     |        10045 | [23, 47]     | KLM(!)              |
|   1 | MAdrid_miLAN     |          NaN | []           | <Air France> (12)   |
|   2 | londON_StockhOlm |        10065 | [24, 43, 87] | (British Airways. ) |
|   3 | Budapest_PaRis   |          NaN | [13]         | 12. Air France      |
|   4 | Brussels_londOn  |        10085 | [67, 32]     | "Swiss Air"         |

(It's some flight data I made up; it's not meant to be accurate in any way.)


38. Some values in the the FlightNumber column are missing (they are NaN). These numbers are meant to increase by 10 with each row so 10055 and 10075 need to be put in place. Modify df to fill in these missing numbers and make the column an integer column (instead of a float column).


In [63]:
df = pd.DataFrame({'From_To': ['LoNDon_paris', 'MAdrid_miLAN', 'londON_StockhOlm','Budapest_PaRis', 'Brussels_londOn'],
'FlightNumber': [10045, np.nan, 10065, np.nan, 10085],'RecentDelays': [[23, 47], [], [24, 43, 87], [13], [67, 32]],'Airline': ['KLM(!)', '<Air France> (12)', '(British Airways. )','12. Air France', '"Swiss Air"']})

In [113]:
df['FlightNumber'] = df['FlightNumber'].fillna(df['FlightNumber'].shift(1) + 10)

In [114]:
df

FlightNumber          Airline      From         To  delay_1  delay_2  \
0       10045.0              KLM    London      paris     23.0     47.0   
1       10055.0       Air France    Madrid      miLAN      NaN      NaN   
2       10065.0  British Airways    London  StockhOlm     24.0     43.0   
3       10075.0       Air France  Budapest      PaRis     13.0      NaN   
4       10085.0        Swiss Air  Brussels     londOn     67.0     32.0   

   delay_3  
0      NaN  
1      NaN  
2     87.0  
3      NaN  
4      NaN

39. The From_To column would be better as two separate columns! Split each string on the underscore delimiter _ to give a new temporary DataFrame called 'temp' with the correct values. Assign the correct column names 'From' and 'To' to this temporary DataFrame.


In [65]:
temp = df.copy()
temp[['From','To']] = temp['From_To'].str.split('_',expand=True)
temp

From_To  FlightNumber  RecentDelays              Airline  \
0      LoNDon_paris       10045.0      [23, 47]               KLM(!)   
1      MAdrid_miLAN           NaN            []    <Air France> (12)   
2  londON_StockhOlm       10065.0  [24, 43, 87]  (British Airways. )   
3    Budapest_PaRis           NaN          [13]       12. Air France   
4   Brussels_londOn       10085.0      [67, 32]          "Swiss Air"   

       From         To  
0    LoNDon      paris  
1    MAdrid      miLAN  
2    londON  StockhOlm  
3  Budapest      PaRis  
4  Brussels     londOn

40. Notice how the capitalisation of the city names is all mixed up in this temporary DataFrame 'temp'. Standardise the strings so that only the first letter is uppercase (e.g. "londON" should become "London".)


In [66]:
temp['From'] = temp['From'].str.capitalize()
temp

From_To  FlightNumber  RecentDelays              Airline  \
0      LoNDon_paris       10045.0      [23, 47]               KLM(!)   
1      MAdrid_miLAN           NaN            []    <Air France> (12)   
2  londON_StockhOlm       10065.0  [24, 43, 87]  (British Airways. )   
3    Budapest_PaRis           NaN          [13]       12. Air France   
4   Brussels_londOn       10085.0      [67, 32]          "Swiss Air"   

       From         To  
0    London      paris  
1    Madrid      miLAN  
2    London  StockhOlm  
3  Budapest      PaRis  
4  Brussels     londOn

41. Delete the From_To column from df and attach the temporary DataFrame 'temp' from the previous questions.


In [67]:
df = df.drop(columns='From_To')
temp = temp.drop(columns='From_To')
df = pd.merge(df, temp, on = ['FlightNumber','Airline'], how = 'inner')
df = df.rename(columns={'RecentDelays_x' : 'RecentDelays'})
df = df.drop(columns='RecentDelays_y')
df

FlightNumber  RecentDelays              Airline      From         To
0       10045.0      [23, 47]               KLM(!)    London      paris
1           NaN            []    <Air France> (12)    Madrid      miLAN
2       10065.0  [24, 43, 87]  (British Airways. )    London  StockhOlm
3           NaN          [13]       12. Air France  Budapest      PaRis
4       10085.0      [67, 32]          "Swiss Air"  Brussels     londOn

42. In the Airline column, you can see some extra puctuation and symbols have appeared around the airline names. Pull out just the airline name. E.g. '(British Airways. )' should become 'British Airways'.


In [70]:
df['Airline'] = df['Airline'].str.replace('([^A-z\s*]+)','',regex=True).str.strip()
df

FlightNumber  RecentDelays          Airline      From         To
0       10045.0      [23, 47]              KLM    London      paris
1           NaN            []       Air France    Madrid      miLAN
2       10065.0  [24, 43, 87]  British Airways    London  StockhOlm
3           NaN          [13]       Air France  Budapest      PaRis
4       10085.0      [67, 32]        Swiss Air  Brussels     londOn

43. In the RecentDelays column, the values have been entered into the DataFrame as a list. We would like each first value in its own column, each second value in its own column, and so on. If there isn't an Nth value, the value should be NaN.

- Expand the Series of lists into a DataFrame named delays, rename the columns delay_1, delay_2, etc. and replace the unwanted RecentDelays column in df with delays.


The DataFrame should look much better now.
| | FlightNumber | Airline | From | To | delay_1 | delay_2 | delay_3 |
|---:|----------------:|:---------------|:---------|:----------|----------:|----------:|----------:|
| 0 | 10045 | KLM | London | Paris | 23 | 47 | NaN |
| 1 | 10055 | Air France | Madrid | Milan | NaN | NaN | NaN |
| 2 | 10065 | British Airways| London | Stockholm | 24 | 43 | 87 |
| 3 | 10075 | Air France | Budapest | Paris | 13 | NaN | NaN |
| 4 | 10085 | Swiss Air | Brussels | London | 67 | 32 | NaN |


In [81]:
# pd.DataFrame(df['RecentDelays'].tolist(), columns = ['delay 1', 'delay 2', 'delay 3'])
df = pd.concat([df.drop(columns='RecentDelays'), df['RecentDelays'].apply(pd.Series).rename(columns=lambda x : f'delay_{x+1}')],axis=1)
df

FlightNumber          Airline      From         To  delay_1  delay_2  \
0       10045.0              KLM    London      paris     23.0     47.0   
1           NaN       Air France    Madrid      miLAN      NaN      NaN   
2       10065.0  British Airways    London  StockhOlm     24.0     43.0   
3           NaN       Air France  Budapest      PaRis     13.0      NaN   
4       10085.0        Swiss Air  Brussels     londOn     67.0     32.0   

   delay_3  
0      NaN  
1      NaN  
2     87.0  
3      NaN  
4      NaN

## Using MultiIndexes

### Go beyond flat DataFrames with additional index levels

#### Difficulty: medium

Previous exercises have seen us analysing data from DataFrames equipped with a single index level. However, pandas also gives you the possibilty of indexing your data using multiple levels. This is very much like adding new dimensions to a Series or a DataFrame. For example, a Series is 1D, but by using a MultiIndex with 2 levels we gain of much the same functionality as a 2D DataFrame.<br>
The set of puzzles below explores how you might use multiple index levels to enhance data analysis.<br>
To warm up, we'll look make a Series with two index levels.


44. Given the lists letters = ['A', 'B', 'C'] and numbers = list(range(10)), construct a MultiIndex object from the product of the two lists. Use it to index a Series of random numbers. Call this Series s.


45. Check the index of s is lexicographically sorted (this is a necessary proprty for indexing to work correctly with a MultiIndex).


46. Select the labels 1, 3 and 6 from the second level of the MultiIndexed Series.


47. Slice the Series s; slice up to label 'B' for the first level and from label 5 onwards for the second level.


48. Sum the values in s for each label in the first level (you should have Series giving you a total for labels A, B and C).


49. Suppose that sum() (and other methods) did not accept a level keyword argument. How else could you perform the equivalent of s.sum(level=1)?


50. Exchange the levels of the MultiIndex so we have an index of the form (letters, numbers). Is this new Series properly lexsorted? If not, sort it.
